[libcxx] Implement C++20 std::chrono::is_clock, std::chrono::is_clock_v

yuxuanchen1997 opened the pull request:
Implemented [[*time.traits.is.clock*]](https://eel.is/c++draft/time.traits.is.clock) from [P0355R7](https://wg21.link/p0355r7).

This patch implements the C++20 feature `is_clock` and `is_clock_v` based on the documentation [on cppreference](https://en.cppreference.com/w/cpp/chrono/is_clock.html)

Fixes #166049.

yuxuanchen1997 committed:
[libcxx] Implement C++20 std::chrono::is_clock, std::chrono::is_clock_v

frederick-vs-ja commented on libcxx/include/chrono:
```diff
@@ -1057,6 +1057,7 @@ constexpr chrono::year                                  operator ""y(unsigned lo
 #    include <__chrono/day.h>
 #    include <__chrono/exception.h>
 #    include <__chrono/hh_mm_ss.h>
+#    include <__chrono/is_clock.h>
```
Yes. I think we should add the parts being implemented.

yuxuanchen1997 commented on libcxx/include/__chrono/is_clock.h:
```diff
@@ -0,0 +1,45 @@
+
+#if _LIBCPP_STD_VER >= 20
+
+#  include <__type_traits/integral_constant.h>
```
The intention was that if the STD_VER was < 20, the transitive include won't happen. Maybe too pessimistic. I can reorder.

philnik777 replied:
The whole header is only included in C++20 and later.

frederick-vs-ja commented on libcxx/include/__chrono/is_clock.h:
```diff
@@ -0,0 +1,72 @@
+
+template <class _Clock, class _Duration, class _ClockType>
+constexpr bool __is_valid_clock_time_point_v<time_point<_Clock, _Duration>, _ClockType> =
+    _IsSame<time_point<_Clock, _Duration>, time_point<_ClockType>>::value ||
```
We can manually add the template argument here:
```suggestion
    _IsSame<time_point<_Clock, _Duration>, time_point<_ClockType, typename _ClockType::duration>>::value ||
```

However, it seems simpler to check the relationship between `_Tp::time_point` and `_Tp::duration`. I.e. `_Tp::time_point` must be of form `time_point<_Clock, typename _Tp::duration>`.

yuxuanchen1997 replied:
How about just change it to 
```
requires _IsSame<typename _Tp::time_point::duration, typename _Tp::duration>::value;
```
?

I think it's the same. Just taking advantage the member type alias.

frederick-vs-ja replied:
Hmm. They're not the same IIUC. We should also check that `_Tp::time_point` is a `std::chrono::time_point` specialization.

frederick-vs-ja commented:
Looks like that CI failures are unrelated. We can update the branch to see whether CI will be green.

frederick-vs-ja reviewed (approved):
LGTM. CC @philnik777

frederick-vs-ja commented:
> > Implemented [[_time.traits.is.clock_]](https://eel.is/c++draft/time.traits.is.clock) from [P0355R7](https://wg21.link/p0355r7).
> > This patch implements the C++20 feature `is_clock` and `is_clock_v` based on the documentation [on cppreference](https://en.cppreference.com/w/cpp/chrono/is_clock.html)
> > It doesn't seem like we track this particular feature in #99982, but it's necessary to implement.
> 
> Would it make sense to create a sub-task to track this item in #99982?

@H-G-Hristov I've just created the corresponding subtask.

@yuxuanchen1997 I modified the PR description to associate this PR with the corresponding subtask. Please double-check.

philnik777 commented on libcxx/include/__chrono/is_clock.h:
```diff
@@ -0,0 +1,73 @@
+class duration;
+
+template <class _Clock, class _Duration>
+class time_point;
```
We shouldn't forward declare classes in random places. Either include the appropriate header or make a forward declaring header.

yuxuanchen1997 replied:
Do you want them to be `__fwd/duration.h` or `__fwd/__chrono/duration.h`?

philnik777 commented on libcxx/include/__chrono/is_clock.h:
```diff
@@ -0,0 +1,73 @@
+  requires __is_ratio_v<typename _Tp::period>;
+
+  typename _Tp::duration;
+  requires _IsSame<typename _Tp::duration, duration<typename _Tp::rep, typename _Tp::period>>::value;
```
```suggestion
  requires same_as<typename _Tp::duration, duration<typename _Tp::rep, typename _Tp::period>>;
```
Same below.

philnik777 commented on libcxx/test/std/time/time.traits.is.clock/trait.is.clock.compile.verify.cpp:
```diff
@@ -0,0 +1,28 @@
+#if !__has_warning("-Winvalid-specializations")
+// expected-no-diagnostics
+#else
+
```
Is this still required? The minimum supported Clang version is 20 now, where the attribute has been implemented already.

frederick-vs-ja commented on libcxx/include/__chrono/is_clock.h:
```diff
@@ -0,0 +1,73 @@
+template <class _Tp>
+_LIBCPP_NO_SPECIALIZATIONS inline constexpr bool is_clock_v = requires {
+  typename _Tp::rep;
+  requires is_arithmetic_v<typename _Tp::rep>;
```
I guess at this moment we can use the following for simplicity. After all, we should submit an LWG issue for clarifying "class emulating an arithmetic type".
```suggestion
  requires is_arithmetic_v<typename _Tp::rep> || is_class_v<typename _Tp::rep> || is_union_v<typename _Tp::rep>;
```
(Note that union types are also class types in the C++ core language.)

yuxuanchen1997 replied:
I was informed by LWG that the issue has been long standing. I was advised, I quote, to "just write the code as though it's an arithmetic type, and if some program-defined class doesn't compile, then that class doesn't emulate an arithmetic type sufficiently closely."

yuxuanchen1997 commented:
@philnik777 @frederick-vs-ja I believe I have addressed all your comments. I am going to land this now. If anything doesn't work feel free to comment back.

Zingam commented on libcxx/test/libcxx/time/time.traits.is.clock/trait.is.clock.compile.verify.cpp:
```diff
@@ -0,0 +1,24 @@
+
+// REQUIRES: std-at-least-c++20
+
+#include <chrono>
```
We usually put a synopsis of what is being tested here. Also the convention is the test name to match the section name + class, e.g. 

libcxx/test/libcxx/time/time.traits.is.clock/trait.is.clock.compile.verify.cpp -> libcxx/test/libcxx/time/time.traits/is.clock.compile.verify.cpp

frederick-vs-ja replied:
Opened #166888 for this.

philnik777 commented:
@yuxuanchen1997 Please wait for approval next time, especially when changes have explicitly been requested. There is no need to rush patches in.

yuxuanchen1997 commented:
> @yuxuanchen1997 Please wait for approval next time, especially when changes have explicitly been requested. There is no need to rush patches in.

Sorry. There was an earlier acceptance and it was my oversight by treating your request to change as an earlier request.
